### PR TITLE
google_grpc: attempt to reduce lock contention between completionThread() and onCompletedOps()

### PR DESCRIPTION
**Pull request by antoniovicente**
Commit Message:
google_grpc: attempt to reduce lock contention between completionThread() and onCompletedOps()

Holding a stream's lock while running handleOpCompletion can result in the completion queue having to wait until the lock is released before adding messages on that stream to completed_ops_.  In cases where the completion queue is shared across multiple gRPC streams, delivery of new messages on all streams is blocked until the lock held by the first stream while executing onCompletedOps.

Additional Description:
Risk Level: low, should be a functional no-op other than reduced lock contention.
Testing: no additional tests yet
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: source/common/grpc/google_async_client_impl.cc
Comment (jmarantz) [inline source/common/grpc/google_async_client_impl.cc:267]: `completed_opts_.clear()` here?

**Review by jmarantz — COMMENTED** [review]
lgtm modulo UB nit (use after move) but I'll let Harvey do the review.

File: source/common/grpc/google_async_client_impl.cc
Comment (htuch) [reply source/common/grpc/google_async_client_impl.cc:267]: Isn't this implicit in move semantics?

**Review by htuch — APPROVED** [review]
LGTM modulo @jmarantz comment resolution. Nice fix.

File: source/common/grpc/google_async_client_impl.cc
Comment (antoniovicente) [reply source/common/grpc/google_async_client_impl.cc:267]: This feels like a repeat of: https://github.com/envoyproxy/envoy/pull/14289/files/606f7af97a41567673a049b609df81cb3b7c99a9#r536419940

Added an ASSERT similar to the one in the dispatcher after a similar std::move.

Comment (jmarantz) [reply source/common/grpc/google_async_client_impl.cc:267]: Oh right; I remember. OK that sounds fine and faster.

**Review by jmarantz — APPROVED** [review]

**Comment by asraa** [comment]
@envoyproxy/senior-maintainers Can a non-googler please approve and merge this?

**Review by ggreenway — APPROVED** [review]

**Pull request merged**
